### PR TITLE
PERF: Merge empty frame

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Faster merge when left or right is empty. 

One ASV added:
```
       before           after         ratio
     [7651c082]       [4f451520]
     <main>           <merge-empty-frame>
-      12.0±0.3ms      1.42±0.09ms     0.12  join_merge.Merge.time_merge_dataframe_empty(False)
-        24.3±2ms      1.29±0.01ms     0.05  join_merge.Merge.time_merge_dataframe_empty(True)
```

Some additional examples:

```
N = 10_000_000

df1 = pd.DataFrame(
    np.random.randint(0, 100, (N, 2)),
    columns=['A', 'B'],
)

df2 = df1.set_index('A')

df_empty = pd.DataFrame(columns=['A', 'C'], dtype='int64')
```

```
%timeit df1.merge(df_empty, how="right")  
242 ms ± 5.32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)        <-- main
1.12 ms ± 50.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- PR
```

```
%timeit df2.merge(df_empty, how="left", left_index=True, right_on='A')    
1.07 s ± 47.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <-- main
241 ms ± 21.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <-- PR
```

### CODE REVIEW

File: asv_bench/benchmarks/join_merge.py [hunk 1]
```diff
@@ -216,6 +216,9 @@ def time_merge_dataframe_integer_2key(self, sort):
     def time_merge_dataframe_integer_key(self, sort):
         merge(self.df, self.df2, on="key1", sort=sort)
 
+    def time_merge_dataframe_empty(self, sort):
```
Comment (jreback) [inline asv_bench/benchmarks/join_merge.py:219]: can you add the reverse as well (e.g. left empty)

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -2055,6 +2066,23 @@ def _get_single_indexer(
     return libjoin.left_outer_join(left_key, right_key, count, sort=sort)
 
 
+def _get_empty_indexer() -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp]]:
+    return (
```
Comment (jreback) [inline pandas/core/reshape/merge.py:2070]: can you add a short doc-string to these both

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -1477,6 +1475,19 @@ def get_join_indexers(
         right_keys
     ), "left_key and right_keys must be the same length"
 
+    left_n = len(left_keys[0])
```
Comment (jreback) [inline pandas/core/reshape/merge.py:1478]: can you add a comment on what you are checking here (e.g. fast-path for empty left/right)

File: pandas/core/reshape/merge.py [hunk 4]
```diff
@@ -1477,6 +1475,19 @@ def get_join_indexers(
...
+    right_n = len(right_keys[0])
+    if left_n == 0:
+        if how in ["left", "inner", "cross"]:
```
Comment (jreback) [inline pandas/core/reshape/merge.py:1481]: do we have tests for all of these empty cases? if not can you add some

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (lukemanley) [reply pandas/core/reshape/merge.py:1479]: comment added

File: pandas/core/reshape/merge.py (hunk 4 above)
Comment (lukemanley) [reply pandas/core/reshape/merge.py:1482]: I think most of these are covered but they are scattered around. I added a parameterized test that covers the different join types and left vs right being empty.

File: pandas/core/reshape/merge.py (hunk 2 above)
Comment (lukemanley) [reply pandas/core/reshape/merge.py:2072]: added docstrings

File: asv_bench/benchmarks/join_merge.py (hunk 1 above)
Comment (lukemanley) [reply asv_bench/benchmarks/join_merge.py]: reverse added here

File: doc/source/whatsnew/v1.5.0.rst [hunk 5]
```diff
@@ -206,6 +206,7 @@ Performance improvements
 - Performance improvement in :meth:`DataFrame.duplicated` when subset consists of only one column (:issue:`45236`)
 - Performance improvement in :meth:`.GroupBy.transform` when broadcasting values for user-defined functions (:issue:`45708`)
 - Performance improvement in :meth:`.GroupBy.transform` for user-defined functions when only a single group exists (:issue:`44977`)
+- Performance improvement in :meth:`DataFrame.merge` when left and/or right are empty (:issue:`45838`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.5.0.rst:209]: I func we normally use :func: merge instead of DataFrame.merge?

Comment (lukemanley) [reply doc/source/whatsnew/v1.5.0.rst]: updated to :func: merge. thanks

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm will merge on green.

tip to avoid conflicts in the release notes; add a note *not* at the bottom, add in middle somewhere

**Comment by phofl** [comment]
Greenish

**Comment by phofl** [comment]
Thx @lukemanley

**Pull request merged**